### ISSUE TITLE
Personal notes about "the software development process"

### ISSUE THREAD

**Issue by laconbass** [laconbass/iai#8]
[Modular Programming on wikipedia]: http://en.wikipedia.org/wiki/Modular_programming
[CIP on wikipedia]: http://en.wikipedia.org/wiki/Continual_improvement_process
[Kaizen on wikipedia]: http://en.wikipedia.org/wiki/Kaizen

> It's clear that those big tasks on *The life cicle* are tasks that "will never "end,
> as part of the ~~continuous~~ continual improvement process that I want to achieve.
>
> _(laconbass @ January 2015)_

This reasoning comes to me while reading about [Modular Programming on wikipedia], while taking a break after #1.

While writting this issue summary, some research has been done about "continual improvement process", some things I should note from now onwards:

1. The word _continual_ should be used instead _continuous_. Why? See [CIP on wikipedia].
2. The japanese word _Kaizen_ arises. This can't be coincidence. See [CIP on wikipedia] and [Kaizen on wikipedia].

My head is also burning out now with posible restructurations of the "goals" and "design principles" documents. A try to conclude why:

> It seems that the "goals" document **should be not _aka. requeriments_**.
> **requeriments seems a thing more specific while goals should be wider.**
>
>  _(laconbass @ January 2015)_

and

> It seems that the "design principles" document is going too wide.
> **architecture design principles should fit on the same place than object oriented design principles?**
> **principles to avoid known blunders should fit on the same place than other design principles?**
>
> _(laconbass @ January 2015)_

I conclude the following tasks should be made before doing anything more:

- [ ] A document describing my personal notes about "the software development process" should be written and maintained.
    - [ ] It should note my deficiencies of knowledge as I'm not an academician of computer science.
    - [ ] It should note my feeling on "how to do things the right way" (continual~~continuous~~ improvement), and maybe the misterious appearing of *Kaizen* while researching to write it down.
- [ ] The whole "Life cycle" document should be refactored because **no clear sequence exists between those big tasks.**
    - [ ] It should reference my personal notes at the first place, as everything for now is depending, and will depend for a while, exclusively on my opinion.
- [ ] Some content from the wiki home "attitude" section should be spliced to other section and:
    - [ ] Explain how *iai* enhances speed (so eficiency) through helping developer have an attitude.
    - [ ] The "attitude" section should state only _why *iai* is considered an attitude_

**Comment by laconbass** [comment]
Created a label to mark on what issue should I focus

**Comment by laconbass** [comment]
This is heavily related to [Software Construction](http://en.wikipedia.org/wiki/Software_construction).

Take a look also at [Software prototyping](http://en.wikipedia.org/wiki/Software_prototyping#Evolutionary_prototyping).

**Comment by laconbass** [comment]
[Modular Programming article on wikipedia]: http://en.wikipedia.org/wiki/Modular_programming
[Mozart-oz documentation]: http://mozart.github.io/documentation/
[Mozart-oz publications]: http://mozart.github.io/publications/
[Programming Paradigms for Dummies: What Every Programmer Should Know]: http://www.info.ucl.ac.be/~pvr/VanRoyChapter.pdf

Related to [Modular Programming article on wikipedia]:

> Modular programming is a software design technique that emphasizes separating the functionality
> of a program into independent, interchangeable modules, such that each contains everything
> necessary to execute only one aspect of the desired functionality.
>
> _([Modular Programming article on wikipedia] @ January 2015)_

This definition is referenced to [Mozart-oz documentation], and leading from there to [Mozart-oz publications] there are lots of stuff that seem valuable for knowledge. My reaction was inmediately to to open Peter Van Roy's [Programming Paradigms for Dummies: What Every Programmer Should Know] , interesting article though.

**Definitions of "module", "package", and "component" should be stablished so things can advance**.

**Comment by laconbass** [comment]
Package is a collection of modules, and each file is a module. A component is a similar concept, but typically refers to a higher level: a component is a piece of a whole system, while a module is a piece of an individual program.

**Comment by lorenzogrv** [comment]
Calling each stage/step/or-the-like PHASE confuses myself. It seems better to think on "the big tasks" as a kind of states or other word that assumes the fact that there isn't a fixed sequence of steps that cycles again and again on the same order (i.e. [:A-B-C-D:]).

The fact is that I jump from any "stage" to any other "stage", so the concept of unknown is here again. It's clear that _working with the unknown_ is a key.

**Comment by lorenzogrv** [comment]
So a kind of a draft of the method/strategy I always apply is:
1. Observe and identify
2. Research and decide
3. Review and evolve (repeating 1, 2 and 3)

Mental verbiage: IDEA => Identify, Decide (or simply "Do"), Evolve, Again. It's not good enough, but keeps the way to the bell ding.

**Comment by laconbass** [comment]
The concept: A mental attitude
- Working trought attitude
- Principles for researching principles
- Avoid known blunders
- The strategy (1 2 3)

**Comment by lorenzogrv** [comment]
Refined the iai concept at the wiki Home page. Think LATER on keeping that knowledge on a separate page, as it is SO IMPORTANT!

"Assumptions" seems a pretty good title for that "personal notes about..." stuff. Start it as a concise introduction for the Life-Cycle document itself, and see what happens.

**Comment by lorenzogrv** [comment]
Done writing those "notes about..." Reworded to "the decision procedure", which actually is what it's need. More tasks:
- [x] Edit the _Sidebar to add links to "concept" and "decision procedure"
- [x] Reference the home on the concept introduction

**Comment by lorenzogrv** [comment]
This issue loses the focus. It has become a self-dicussion about the decision procedure so the job here is done for now (until it needs review). A new issue regarding the life cycle must be created and I'd focus on it.

Note this issue can only be closed when _the decision procedure_ is successfully applied to refactor the life cycle. The last task before close would be:
- [ ] Refactor this issue summary

**Comment by laconbass** [comment]
The last task before close would be:

- [ ] Refactor this issue summary

**Comment by lorenzogrv** [comment]
When discussing and researching it probably belongs to an issue
